4.0.0

Daniele-rolli opened the pull request:
## Pull Request Type
- [x] **Feature/Component** (New feature or component)
- [x] **Bug Fix** (Fixes an issue or bug)

## Checklist (For Features/Bug Fixes)

- [x] Does this PR address the **intended feature/bug** without any unintended changes?
- [x] Have all **new components/features** been tested for functionality?
- [x] Have all **bugs been fixed** and verified?
- [x] Is the **UI/UX** consistent across different screen/window sizes?
- [x] Have **translations** been updated for all new user-facing components?
- [x] Have you confirmed that the **same work has been applied / will be applied to the mobile app** (if applicable)?
- [x] Is this PR ready for release, with appropriate versioning?

## Changelog:

1. Fixed issue #285
2. Fixed issue #287
3. Improved collapsible headings
4. Overhauled sync functionality
5. Fixed issue #282
6. Enhanced highlighters
7. Upgraded from Electron 31 to 35
8. Fixed media issues in Electron 35
9. Improved recording (styling, and pause/resume buttons)
10. Enhanced sharing (.bea notes now open upon import)
11. Added logic to dynamically import translations in `index.js`
12. Fixed issue #262 
13. Fixed issue #292 
14. Fixed issue #291

bigshans commented:
Hi @Daniele-rolli , now creating a new note by shortcut will get two new notes. It seems to be caused by duplicate definitions of menu shortcuts and web page shortcuts.

Daniele-rolli commented:
I've fixed the issue, but I'm still deep into working on cloud sync for Pocket. Could you look into issue #287 before the release? Thanks!

Daniele-rolli commented:
No need to look into #287, it was a pretty straightforward fix. Just give it a look and let me know if you spot any issues in the paste logic

Daniele-rolli commented:
I discovered some issues in the CSS styling that were causing problems with paragraph spacing. It's fixed now, but many notes might still have some issues because the spacing was previously handled by:

```html
<p> hello </p>
<p><br class="ProseMirror-trailingBreak"></p>
<p></p>
```
Now, with the proper spacing, the `<p><br class="ProseMirror-trailingBreak"></p>` element sticks out. This has been changed because, before, paragraphs were copied with three line breaks.

bigshans commented:
One bug about paste:

https://github.com/user-attachments/assets/ea63e5a3-f18b-4eed-899f-e10cc46ed82b

Daniele-rolli commented:
I'm in a bit of a time crunch. The paste logic is still buggy, if you could look into it for me, I'd really appreciate it.

Daniele-rolli commented:
I've made a couple of changes. Give it a spin and let me know if everything seems fine. Translations still need some work.

bigshans commented:
It would be better to add an option to clear the highlight.

Daniele-rolli commented:
Clicking the color already does that, and it's implemented the same way in many other apps.

bigshans commented:
> Clicking the color already does that, and it's implemented the same way in many other apps.

Yes, I know. However, the highlight color of the currently selected item is not distinguishable from other colors, which makes it difficult to differentiate for users with color vision deficiency.

Also, is it just me, or does the highlight color shown here seem to vary across different themes?

![图片](https://github.com/user-attachments/assets/68c2e2d8-010d-4727-99f5-943753e21c79)
![图片](https://github.com/user-attachments/assets/28660d1f-6a76-4c23-aafb-616f7954c854)

Daniele-rolli commented:
The colors vary across different themes to improve legibility. The two rows above were added to provide more standard, non pastel, colors while the two below are the original color rows we introduced a few releases ago. Of course, I'm open to changes.

bigshans commented:
It's just a bit strange. In dark mode, the styles of the upper and lower sections feel quite disconnected. To me, the top two lines appear too dim, while the bottom two are overly bright. Moreover, the color difference between the two lines at the bottom isn't very noticeable, which makes it a bit difficult for me to tell them apart.

Daniele-rolli commented:
I get it! Here's my fix, since some colors were redundant and confusing, I rewrote the palette to match the one above. Try it and let me know what you think.

bigshans commented:
I think it's OK.

Daniele-rolli commented:
I've updated the remaining translations. I'll release them as soon as I get the all clear from you.

Daniele-rolli commented:
Hey @bigshans,
When you have a moment, could you give this a spin? That way, we can move forward with releasing this version. Just a heads up, this PR includes two breaking changes:
- A new sync handling method
- Fixed padding in the editor (which might affect formatting)

No worries, though! I’ve added a fix so that if you modify a note and then reopen it, any extra white space should get cleaned up automatically. Thanks a bunch!

bigshans commented:
In my opinion, if auto sync is an option, then there should be a manual sync to go with it.

And I found a very weird bug which not been shown before.

https://github.com/user-attachments/assets/772a3acc-1fce-4483-a254-2647b0fc494b

bigshans commented:
Well, if the user doesn't enable auto sync, how can they sync the data currently? Although import and export are also a way. About git sync, I think it's another matter. One is the sync method, and the another one is the sync entry.

Daniele-rolli commented:
The sync code includes logic to force a sync, so we could leverage that by adding a button to manually trigger a sync, even when auto-sync is disabled. This way, users still get the benefits of cloud sync through manual input. What do you think?

bigshans commented:
> The sync code includes logic to force a sync, so we could leverage that by adding a button to manually trigger a sync, even when auto-sync is disabled. This way, users still get the benefits of cloud sync through manual input. What do you think?

Yes, this is what I mean.

bigshans commented:
I reproduce the bug on my Windows again. You need set interface size as Large Medium.

Daniele-rolli commented:
I’m still having no luck replicating this, no matter what I try! Any updates on your end?

https://github.com/user-attachments/assets/217c8249-076b-4efc-ba9b-6cf4d301ee2c

bigshans commented:
Move the cursor to the settings gear and then roll the wheel.

https://github.com/user-attachments/assets/63961c5f-3b1d-4727-b18d-ec8b18249063

Daniele-rolli commented:
Found the issue, it’s related to Tippy. Basically, page zoom messes up the tooltip position. I’m currently trying to find a fix. Both [this issue on tippyjs-react](https://github.com/atomiks/tippyjs-react/issues/80) and [this one on floating-ui](https://github.com/floating-ui/floating-ui/issues/3032) didn’t help.
